### PR TITLE
[CDAP-14882] Adding log buffer processor pipeline

### PR DESCRIPTION
**Pull request by CuriousVini**
First commit:
Adding log buffer processor pipeline. The pipeline will receive logs from `ConcurrentLogBufferWriter` and those logs will be enqueued in `BlockingQueue`. The pipeline will process these events and send them to appenders.

Second commit:

* Migrated log checkpoint table to use new apis.
* Added `AbstractCheckpointManager` which stores checkpoints with following schema
  String row_prefix, int partition and  bytes checkpoint. The concrete implementations of this class are `KafkaCheckpointManager` and `LogBufferCheckpoint` manager which provides methods to serialize and deserialize checkpoints.

Tested changes on hadoop cluster.

DUT: https://builds.cask.co/browse/CDAP-DUT6833-5
ITM: https://builds.cask.co/browse/IT-ITM-104

### CODE REVIEW

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/logbuffer/ConcurrentLogBufferWriter.java [hunk 1]
```diff
@@ -167,6 +167,8 @@ void process(LogBufferWriter writer) {
         // after log events have been sent to log buffer pipeline for processing, set the pending requests to complete
         completeAll(null);
       } catch (Throwable t) {
+        // This will only happen if logs were not written to log buffer(WAL), so the log buffer request must be
+        // retried. So complete all the requests with error.
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/logbuffer/ConcurrentLogBufferWriter.java:171]: nice comment

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+
+    // none of the events were processed.
+    if (metadata.getTotalEventsProcessed() <= 0) {
+      return false;
+    }
+
+    unSyncedEvents += metadata.getTotalEventsProcessed();
+    // events were processed, so update the checkpoints
+    Checkpoint<LogBufferFileOffset> checkpoint = metadata.getCheckpoints().get(0);
+    MutableLogBufferCheckpoint mutableCheckpoint = checkpoints.get(0);
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java]: I'm a bit confused about this map, there is only ever one item in it right? Is it used to make it conform to the checkpointManager API? Are the keys supposed to be partition numbers or something like that?

In any case, I think it's more clear if we also make the checkpoint a variable, and just use the map when passing it to the checkpoint manager.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+  /**
+   * Persists the checkpoints.
+   */
+  private void persistCheckpoints() {
+    try {
+      checkpointManager.saveCheckpoints(checkpoints);
+      LOG.debug("Checkpoint persisted for {} with {}", name, checkpoints);
+    } catch (Exception e) {
+      // Just log as it is non-fatal if failed to save checkpoints
+      OUTAGE_LOG.warn("Failed to persist checkpoint for pipeline {}.", name, e);
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java:197]: Are we sure we want to log and continue if the checkpoint can't be saved? What types of exceptions can be thrown when saving checkpoints? Do we want to retry?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+   * Sync the appender and persists checkpoints if it is time.
+   *
+   * @return delay in millisecond till the next sync time.
+   */
+  private long trySyncAndPersistCheckpoints(long currentTimeMillis) {
+    if (unSyncedEvents == 0) {
+      return config.getCheckpointIntervalMillis();
+    }
+    if (currentTimeMillis - config.getCheckpointIntervalMillis() < lastCheckpointTime) {
+      return config.getCheckpointIntervalMillis() - currentTimeMillis + lastCheckpointTime;
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java:211]: nit: I think it's easier to understand if this is lastCheckpointTime + config.getCheckpointIntervalMillis() - currentTimeMillis

It was confusing to me at first why we're subtracting the current timestamp from the interval.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+    try {
+      context.sync();
+      // Only persist if sync succeeded. Since persistCheckpoints never throw, it's ok to be inside the try.
+      persistCheckpoints();
+      lastCheckpointTime = currentTimeMillis;
+      metricsContext.gauge("last.checkpoint.time", lastCheckpointTime);
+      unSyncedEvents = 0;
+      LOG.debug("Events synced and checkpoint persisted for {}", name);
+    } catch (Exception e) {
+      OUTAGE_LOG.warn("Failed to sync in pipeline {}. Will be retried.", name, e);
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java:224]: What does the retrying?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+    private long filePos;
+
+    MutableLogBufferFileOffset(long fileId, long filePos) {
+      super(fileId, filePos);
+      this.fileId = fileId;
+      this.filePos = filePos;
+    }
+
+    @Override
+    public long getFileId() {
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java:279]: if everything here is private, can access the fields directly rather than through getters and setters.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+        removeProcessedEvent();
+        return endOfData();
+      }
+
+      if (queue.peek() == null) {
+        return endOfData();
+      }
+
+      // Remove already processed event from the queue
+      removeProcessedEvent();
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java]: Not sure I understand the logic in this class. It seems like it would be simpler to remove prevLogEvent, poll here, increment, and return a new ProcessorEvent. How come we need to remove, then peek?

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/pipeline/LogPipelineTestUtil.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    event.setMessage(message);
+    event.setTimeStamp(timestamp);
+    return event;
+  }
+
+  /**
+   * Creates logger context.
+   */
+  public static LoggerContext createLoggerContext(String rootLevel,
+                                            Map<String, String> loggerLevels,
```
Comment (albertshau) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/pipeline/LogPipelineTestUtil.java]: nit: indentation

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java (hunk 6 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java:279]: This is for checkpoint manager to access mutated state.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java (hunk 7 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java]: The time event queue processor may not process the event if underlying queue is already full https://github.com/cdapio/cdap/blob/develop/cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/queue/TimeEventQueueProcessor.java#L82 
When `iterator.hasNext` is called from time event queue processor, that call will indicate that  previous event has been processed. So we can safely remove `prevLogEvent` and then peek next event for further processing.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java (hunk 5 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java:224]: It will be done by run method https://github.com/cdapio/cdap/pull/11080/files#diff-0def6fa6acd833eaa7edbbb45acf188dR103

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java (hunk 3 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java:197]: If the checkpointing failed when called from `trySyncAndPersistCheckpoints` method, it will be retried again by run method in next iteration. However, if the its called from shutdown method then it won't be retried.

**Comment by CuriousVini** [comment]
Addressing comments and fixing merge conflict on `StoreDefinition.java` file

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/LogPipelineLoader.java [hunk 9]
```diff
@@ -58,11 +56,8 @@
   private static final Logger LOG = LoggerFactory.getLogger(LogPipelineLoader.class);
   private static final String SYSTEM_LOG_PIPELINE_CONFIG = "cdap-log-pipeline.xml";
   private static final String SYSTEM_LOG_PIPELINE_NAME = "cdap";
-  private static final Set<byte[]> RESERVED_CHECKPOINT_PREFIX = ImmutableSortedSet.orderedBy(Bytes.BYTES_COMPARATOR)
-    .add(Bytes.toBytes(101))    // This is used by the old metrics log plugin
-    .add(Bytes.toBytes(200))    // This is used by the old file meta data
-    .add(Bytes.toBytes(300))    // This is used by the new file meta data
-    .build();
+  // This is used by the new file meta data
+  private static final Set<String> RESERVED_CHECKPOINT_PREFIX = ImmutableSortedSet.of("log_metadata");
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/LogPipelineLoader.java]: can this just be a String?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java (hunk 7 above)
Comment (albertshau) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java]: How is that related to how this is implemented? Is the queue being read from somewhere else? Can't this be implemented like:

```
if (count >= batchSize) {
  return endOfData();
}

LogBufferEvent event = queue.poll();
if (event == null) {
  return endOfData();
}
count++;
return new ProcessorEvent<>(...);
```

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java (hunk 6 above)
Comment (albertshau) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java:279]: oh I see, these are overrides.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java (hunk 5 above)
Comment (albertshau) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java:224]: In that case, it's better if this method throws the exception and the logging/handling is done by the caller that is actually doing the retrying.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java (hunk 3 above)
Comment (albertshau) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java:197]: But is it correct to continue on in that scenario? For example, if the checkpoint store is down for some time, that means it will keep syncing log events even though the checkpoint is far behind. What happens if the log service then dies? Will it rewrite all those logs again?

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 10]
```diff
@@ -706,4 +711,27 @@ public static void createTables(StructuredTableAdmin tableAdmin) throws IOExcept
       tableAdmin.create(JOB_QUEUE_STORE_SPEC);
     }
   }
+
+  /**
+   * Schema for log checkpoint store.
+   */
+  public static final class LogCheckpointStore {
+    public static final StructuredTableId LOG_CHECKPOINT_TABLE = new StructuredTableId("log_checkpoints");
+    public static final String ROW_PREFIX_FIELD = "row_prefix";
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: just 'prefix' seems like a better name

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+  @VisibleForTesting
+  public AbstractCheckpointManager(TransactionRunner transactionRunner, String prefix) {
+    this.rowKeyPrefix = prefix;
+    this.lastCheckpoint = new HashMap<>();
+    this.transactionRunner = transactionRunner;
+  }
+
+  @Override
+  public void saveCheckpoints(final Map<Integer, ? extends Checkpoint<T>> checkpoints) throws IOException {
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java]: does checkpoints need to be final?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+    LOG.trace("Saved checkpoints for partitions {}", checkpoints);
+  }
+
+  @Override
+  public Map<Integer, Checkpoint<T>> getCheckpoint(final Set<Integer> partitions) throws IOException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      Map<Integer, Checkpoint<T>> checkpoints = new HashMap<>();
+      StructuredTable table = context.getTable(StoreDefinition.LogCheckpointStore.LOG_CHECKPOINT_TABLE);
+      for (final int partition : partitions) {
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java]: doesn't need to be final

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+      }
+      return result;
+
+    }, IOException.class);
+
+    LOG.trace("Saved checkpoints for partitions {}", checkpoints);
+  }
+
+  @Override
+  public Map<Integer, Checkpoint<T>> getCheckpoint(final Set<Integer> partitions) throws IOException {
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java]: doesn't need to be final

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+   * Deserialize the checkpoint bytes. If checkpoint bytes are null, then return {@code -1} for all the numeric
+   * values of checkpoint such as maxEventTime.
+   * @param checkpoint checkpoint bytes to be deserialized
+   * @return deserialized checkpoint
+   * @throws IOException if error while deserializing checkpoint bytes
+   */
+  protected abstract Checkpoint<T> deserializeCheckpoint(@Nullable byte[] checkpoint) throws IOException;
+
+  private Collection<Field<?>> getKeyFields(String rowPrefix, int partition) {
+    return Arrays.asList(Fields.stringField(StoreDefinition.LogCheckpointStore.ROW_PREFIX_FIELD, rowPrefix),
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java]: it's a bit more performant to create a List and add the fields directly instead of converting an array into a list.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+      }
+      return checkpoints;
+    }, IOException.class);
+  }
+
+  @Override
+  public Checkpoint<T> getCheckpoint(final int partition) throws IOException {
+    Checkpoint<T> checkpoint = TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(StoreDefinition.LogCheckpointStore.LOG_CHECKPOINT_TABLE);
+      List<Field<?>> keyFields = ImmutableList.<Field<?>>builder()
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java]: can just use the result of getKeyFields(), it doesn't have to be immutable.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+    LOG.trace("Saved checkpoints for partitions {}", checkpoints);
+  }
+
+  @Override
+  public Map<Integer, Checkpoint<T>> getCheckpoint(final Set<Integer> partitions) throws IOException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      Map<Integer, Checkpoint<T>> checkpoints = new HashMap<>();
+      StructuredTable table = context.getTable(StoreDefinition.LogCheckpointStore.LOG_CHECKPOINT_TABLE);
+      for (final int partition : partitions) {
+        List<Field<?>> keyFields = ImmutableList.<Field<?>>builder()
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java]: can just use the result of getKeyFields, it doesn't need to be immutable

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+    if (lastCheckpoint.equals(checkpoints)) {
+      return;
+    }
+
+    lastCheckpoint = TransactionRunners.run(transactionRunner, context -> {
+      Map<Integer, Checkpoint<T>> result = new HashMap<>();
+      StructuredTable table = context.getTable(StoreDefinition.LogCheckpointStore.LOG_CHECKPOINT_TABLE);
+      for (Map.Entry<Integer, ? extends Checkpoint<T>> entry : checkpoints.entrySet()) {
+        Checkpoint<T> checkpoint = entry.getValue();
+        List<Field<?>> fields = ImmutableList.<Field<?>>builder()
```
Comment (albertshau) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java]: if getKeyFields() returns a mutable list, you can just add the checkpoint field to the key fields. That would be a little faster than making an immutable copy.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java (hunk 11 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java]: This is same as what we had in DefaultCheckpointManager: https://github.com/cdapio/cdap/blob/develop/cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/DefaultCheckpointManager.java#L105
removed final

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java (hunk 7 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java]: Ok, had an offline discussion with @albertshau about this. Since the AbstractIterator makes the implementation complicated, using `Iterator` is much simpler. In that case, iterator does not keep to keep track of `prevEvent`.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java (hunk 3 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java:197]: Log saver makes sure there is no loss of events in exception cases. However, there may be log duplication. Added comment to clarify that.

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by CuriousVini** [comment]
Fixed merge conflicts with StoreDefinition class and squashed all the commits.

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 18]
```diff
@@ -715,6 +720,7 @@ public static void createTables(StructuredTableAdmin tableAdmin) throws IOExcept
   }
 
   /**
+<<<<<<< HEAD
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: ?? Merge conflict?

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 19]
```diff
@@ -16,6 +16,7 @@
 
 package co.cask.cdap.store;
 
+import co.cask.cdap.data2.metadata.lineage.LineageTable;
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: Why is this added?

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java (hunk 18 above)
Comment (CuriousVini) [reply cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: oops..Fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+  /**
+   * Serialize the checkpoint.
+   * @param checkpoint checkpoint to be serialized
+   * @return serialized checkpoint byte array
+   * @throws IOException if error while serializing checkpoint
+   */
+  protected abstract byte[] serializeCheckpoint(Checkpoint<T> checkpoint) throws IOException;
+
+  /**
+   * Deserialize the checkpoint bytes. If checkpoint bytes are null, then return {@code -1} for all the numeric
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java]: Why require returning `-1` ? Seems like shouldn't be a contract on an abstract method. It is implementation specific to return a `Checkpoint` object that represents the initial state for the corresponding implementation.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/CheckpointManager.java [hunk 21]
```diff
@@ -28,17 +29,17 @@
...
-  void saveCheckpoints(Map<Integer, ? extends Checkpoint<Offset>> checkpoints) throws Exception;
+  void saveCheckpoints(Map<Integer, ? extends Checkpoint<Offset>> checkpoints) throws IOException;
 
   /**
    * Reads the set of {@link Checkpoint}s for the given set of partitions. If there is no checkpoint for the partition,
-   * a {@link Checkpoint} with both {@link Checkpoint#getMaxEventTime()} returning {@code -1} will be used.
+   * {@code -1} value will be used for maxEventTime and numeric Offset values.
    */
-  Map<Integer, Checkpoint<Offset>> getCheckpoint(Set<Integer> partitions) throws Exception;
+  Map<Integer, Checkpoint<Offset>> getCheckpoint(Set<Integer> partitions) throws IOException;
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/CheckpointManager.java]: The `Offset` is a type parameter, right? It should be all caps as `OFFSET`.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaLogProcessorPipeline.java [hunk 22]
```diff
@@ -361,7 +361,7 @@ private long trySyncAndPersistCheckpoints(long currentTimeMillis) {
       return config.getCheckpointIntervalMillis();
     }
     if (currentTimeMillis - config.getCheckpointIntervalMillis() < lastCheckpointTime) {
-      return config.getCheckpointIntervalMillis() - currentTimeMillis + lastCheckpointTime;
+      return lastCheckpointTime + config.getCheckpointIntervalMillis() - currentTimeMillis;
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaLogProcessorPipeline.java:364]: Why change the order?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java [hunk 23]
```diff
@@ -0,0 +1,347 @@
...
+    } catch (InterruptedException e) {
+      // Interruption means stopping the service.
+    }
+  }
+
+  @Override
+  protected void triggerShutdown() {
+    stopped = true;
+    if (runThread != null) {
+      runThread.interrupt();
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java]: Better don't use interrupt.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java [hunk 24]
```diff
@@ -0,0 +1,347 @@
...
+  }
+
+  /**
+   * This method is called from {@link ConcurrentLogBufferWriter} to push log events.
+   *
+   * @param events log events to be processed
+   */
+  public void processLogEvents(Iterator<LogBufferEvent> events) {
+    // Don't accept any log events if the pipeline is not running
+    if (!isRunning()) {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java]: Better use the `stopped` flag. Also, I think the check should be part of the `while` loop condition.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java [hunk 25]
```diff
@@ -0,0 +1,347 @@
...
+        long now = System.currentTimeMillis();
+        long nextCheckpointDelay = trySyncAndPersistCheckpoints(now);
+
+        // If nothing has been processed (e.g. fail to append anything to appender),
+        // Sleep until min(next checkpoint delay, next event delay).
+        if (!hasEventProcessed) {
+          long sleepMillis = config.getEventDelayMillis();
+          sleepMillis = Math.min(sleepMillis, nextCheckpointDelay);
+          if (sleepMillis > 0) {
+            TimeUnit.MILLISECONDS.sleep(sleepMillis);
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java]: Instead of sleep (which requires an interrupt to unblock it), use a `CountDownLatch` as a stop latch, which will get count down when the `triggerShutdown` is called. Then in here, you can do a blocking sleep as `stopLatch.await(sleepMillis, TimeUnit.MILLISECONDS);`

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java [hunk 26]
```diff
@@ -0,0 +1,347 @@
...
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.debug("Shutting down log processor pipeline for {}", name);
+
+    try {
+      context.stop();
+      // Persist the checkpoints. It can only be done after successfully stopping the appenders.
+      // Since persistCheckpoint never throw, putting it inside try is ok.
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java]: then why not move it outside of the try??

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java [hunk 27]
```diff
@@ -0,0 +1,347 @@
...
+    MutableLogBufferFileOffset offset = mutableCheckpoint.getOffset();
+    offset.setFileId(checkpoint.getOffset().getFileId());
+    offset.setFilePos(checkpoint.getOffset().getFilePos());
+    mutableCheckpoint.setMaxEventTs(checkpoint.getMaxEventTime());
+
+    return true;
+  }
+
+  /**
+   * This method is called from {@link ConcurrentLogBufferWriter} to push log events.
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java]: Shouldn't need to mention the `ConcurrentLogBufferWriter`. It can be called from anywhere, right?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/queue/TimeEventQueueProcessor.java [hunk 28]
```diff
@@ -89,6 +89,7 @@ public TimeEventQueueProcessor(LogProcessorPipelineContext context, long maxBuff
       ProcessorEvent<Offset> processorEvent = eventIterator.next();
       eventQueue.add(processorEvent.getEvent(), processorEvent.getEvent().getTimeStamp(), processorEvent.getEventSize(),
                      partition, processorEvent.getOffset());
+
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/queue/TimeEventQueueProcessor.java]: Unnecessary new line

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java (hunk 19 above)
Comment (CuriousVini) [reply cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: This was to fix the javadoc https://github.com/cdapio/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java#L635

I guess we should keep it consistent with other javadocs. fixed

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaLogProcessorPipeline.java (hunk 22 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaLogProcessorPipeline.java:364]: this is as per https://github.com/cdapio/cdap/pull/11080#discussion_r258616757

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/CheckpointManager.java (hunk 21 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/CheckpointManager.java]: Followed pattern from https://github.com/cdapio/cdap/blob/develop/cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/queue/TimeEventQueue.java#L34

Will change it.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java (hunk 26 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java]: This is same as KafkaLogProcessorPipeline. But changed as per the comment

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/CheckpointManager.java (hunk 21 above)
Comment (chtyim) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/CheckpointManager.java]: All parameterized type must be all caps, to avoid confusion with a class name.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java (hunk 20 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java]: This is because `CheckpointManager` expects -1 as value if checkpoint is missing: https://github.com/cdapio/cdap/blob/develop/cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/CheckpointManager.java#L35

Comment (chtyim) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java]: And then why the `CheckpointManager` requires `-1`?

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java (hunk 25 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java]: nice! Fixed.

File: cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java (hunk 20 above)
Comment (CuriousVini) [reply cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java]: Removed from contract so that implementation can choose what should be the initial state of the checkpoint.

**Review by chtyim — APPROVED** [review]
LGTM

**Comment by CuriousVini** [comment]
Fixed merge conflicts in `StoreDefinition.java`, running final builds:
DUT: https://builds.cask.co/browse/CDAP-DUT6833-10
ITM: https://builds.cask.co/browse/IT-ITM-113

**Comment by CuriousVini** [comment]
tested changes on hadoop cluster.Both builds are green. Squashing last commit and merging.
